File: 3 lab/g.cpp
// #include <iostream>
// #include <algorithm>
// using namespace std;



// bool canFly(int mid, int arr[], int n, int f) {
//     int cnt = 0;
//     for (int i = 0; i < n; i++) {
//         cnt += (arr[i] + mid - 1) / mid;  
//         if (cnt > f) {
//             return false;
//         }
//     }
//     return true;
// }

// int main() {
//     int n, f;
//     cin >> n >> f;
//     int arr[n];
//     for (int i = 0; i < n; i++) {
//         cin>>arr[i];
//     }
//     sort(arr, arr+n);
//     int l = 0, r = arr[n-1], ans = -1;
//     while (l<=r){
//         int mid = l + (r-l)/2;
//         if(canFly(mid, arr, n, f)){
//             ans = mid;
//             r = mid -1;
//         }
//         else{
//             l = mid + 1;
//         }
        
//     }
//     cout<<ans;

//     return 0;
// }

#include <iostream>
#include <algorithm>
using namespace std;

bool canFly(int mid, int arr[], int n, int f) {
    int cnt = 0;
    for (int i = 0; i < n; i++) {
        cnt += (arr[i] + mid - 1) / mid; 
        if (cnt > f) {
            return false;
        }
    }
    return true;
}

int main() {
    int n, f;
    cin >> n >> f;
    int arr[n];
    int totalChildren = 0;  
    for (int i = 0; i < n; i++) {
        cin >> arr[i];
        totalChildren += arr[i];
    }
    int l = 1, r = totalChildren, ans = -1;  
    while (l <= r) {
        int mid = l + (r - l) / 2;
        if (canFly(mid, arr, n, f)) {
            ans = mid;
            r = mid - 1;
        } else {
            l = mid + 1;
        }
    }
    cout << ans;

    return 0;
}

File: 3 lab/g.py
def canFly(mid, arr, n, f):
    cnt = 0
    for i in range(n):
        temp = arr[i]
        while temp > 0:
            temp -= mid
            cnt += 1
        if cnt > f:
            break
        cnt = 0
    return cnt <= f


def main():
    # n, f = map(int, input().split())
    # arr = list(map(int, input().split()))
    n = 3
    f = 6
    arr = [10, 10, 10]
    arr.sort()
    l, r = 0, arr[-1]
    ans = -1
    while l <= r:
        mid = l + (r - l) // 2
        if canFly(mid, arr.copy(), n, f):
            ans = mid
            r = mid - 1
        else:
            l = mid + 1
    print(ans)


if __name__ == "__main__":
    main()
